README.md
# GerNameR
Collection of names and associations to names for German

## Usage

### Installation

The software can be installed directly from github using the devtools package. Just type

```R
devtools::install_github("aggloeckner/GerNameR")
```

at the R prompt to install the software. If you don't have devtools installed currently, please install the devtools package via the
command

```R
install.packages("devtools")
```

### Selecting Names

Names can be selected and grouped using a variety of commands. The simples method is to filter based on attributes.
You can use the method "filter.names()" for this. E.g. 
```R
filter.names(Intelligence > 0.8)
```
will select all names where Intelligence is above the 80th percentile.

### Grouping names

To create stimulus materials it is often necesary to group names into pairs of most similar names. To do this, select 
sets to be paired (if you do not have any criterion to split the names on, e.g. to compare male and female names,
you can select random sets) and perform the grouping using "match.split()". E.g.
```
# Create a split based on Sex
s <- partition.names(Sex)

# Match pairs of male and female names
m <- match.split( s )
```
will create two groups of male and female names and then create pairs of names with each pair consisting of one male
and one female name.


NEWS
GerNameR v0.3.0 (Release date: 2018-12-14)
==============

* Updated dataset with additional measurements

GerNameR v0.2.1 (Release date: 2018-07-09)
==============

* Added templates for ISSUES
* Added README
* Added code of conduct

GerNameR v0.2.0 (Release date: 2018-07-06)
==============

* Added methods to retrieve ratings for names
* Added methods to pair names from separate subsets (not only median split)
* Added methods to convert pairs and subsets to data frames
* Added method to randomly split sets of names

GerNameR v0.1.2 (Release date: 2018-07-03)
==============

* Updated code for rlang version 0.2.0

GerNameR v0.1.1 (Release date: 2018-23-03)
==============

* Included URL in description
* Updated title

GerNameR v0.1.0 (Release date: 2018-19-03)
==============

* Initial versions
* Features:
  * Selection of subsets of names
  * Distance calculation between sets of names
  * Pairing of names (trial generation)
  * Creation of similar groups of names


tests/testthat.R
# Generic test driver for testthat package
#
# Copyright (C) 2016 Tillmann Nett for FernUni Hagen
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
###############################################################################

library(testthat)
library(GerNameR)

test_check("GerNameR")


tests/testthat/test_match.R
# Copyright (C) 2018 Tillmann Nett for FernUni Hagen
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
###############################################################################

context("Matching")

###############################################

test_that("Matches have correct size", {

  ##### Setup

  s0 <- filter.names()
  s1 <- filter.names(Familiarity >= 0.5, Nationality >= 0.5)
  sp1 <- partition.names(Sex)
  sp2 <- partition.names(Sex, discard = 0.2)
  sp3 <- partition.names(Sex, discard = 0.2, subset = s1)
  sp4 <- partition.names.random()
  sp5 <- partition.names.random(subset = s1)
  sp6 <- partition.names.random(prop=c(1,2))
  sp7 <- partition.names.random(subset = s1, prop=c(1,2))
  sp8 <- partition.names.random(prop=c(1,2,4))

  ##### Match with split based on attribute

  m1 <- match.partition(Sex)

  # We should not have more than original dataset
  expect_true( (length(m1[1]) + length(m1[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m1[1]), length(m1[2]))


  m2 <- match.partition(Sex, subset=s0)

  # We should not have more than original dataset
  expect_true( (length(m2[1]) + length(m2[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m2[1]), length(m2[2]))

  m3 <- match.partition(Sex, subset=s1)

  # We should not have more than original dataset
  expect_true( (length(m3[1]) + length(m3[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m3[1]), length(m3[2]))


  m4 <- match.partition(Sex, discard = 0.2)

  # We should not have more than original dataset
  expect_true( (length(m4[1]) + length(m4[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m4[1]), length(m4[2]))

  m5 <- match.partition(Sex, discard = 0.2, subset=s0)

  # We should not have more than original dataset
  expect_true( (length(m5[1]) + length(m5[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m5[1]), length(m5[2]))

  m6 <- match.partition(Sex, discard = 0.2, subset=s1)

  # We should not have more than original dataset
  expect_true( (length(m6[1]) + length(m6[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m6[1]), length(m6[2]))

  m7 <- match.partition(Sex, discard = 0.2, subset=s0, Competence=10, Intelligence=10)

  # We should not have more than original dataset
  expect_true( (length(m7[1]) + length(m7[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m7[1]), length(m7[2]))


  m8 <- match.partition(Sex, discard = 0.2, subset=s1, Competence=10, Intelligence=10)

  # We should not have more than original dataset
  expect_true( (length(m8[1]) + length(m8[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m8[1]), length(m8[2]))

  ##### Match based on predefined split

  m9 <- match.split(sp1)
  m10 <- match.split(sp1, Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m9[1]) + length(m9[2])) <= length(s0))
  expect_true( (length(m10[1]) + length(m10[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m9[1]), length(m9[2]))
  expect_equal(length(m10[1]), length(m10[2]))
  # We should match as many names as possible
  expect_equal(min(length(m9[1]),length(m9[2])),   min(length(sp1[1]),length(sp1[2])) )
  expect_equal(min(length(m10[1]),length(m10[2])), min(length(sp1[1]),length(sp1[2])) )


  m11 <- match.split(sp2)
  m12 <- match.split(sp2, Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m11[1]) + length(m11[2])) <= length(s0))
  expect_true( (length(m12[1]) + length(m12[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m11[1]), length(m11[2]))
  expect_equal(length(m12[1]), length(m12[2]))
  # We should match as many names as possible
  expect_equal(min(length(m11[1]),length(m11[2])), min(length(sp2[1]),length(sp2[2])) )
  expect_equal(min(length(m12[1]),length(m12[2])), min(length(sp2[1]),length(sp2[2])) )


  m13 <- match.split(sp3)
  m14 <- match.split(sp3, Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m13[1]) + length(m13[2])) <= length(s1))
  expect_true( (length(m14[1]) + length(m14[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m13[1]), length(m13[2]))
  expect_equal(length(m14[1]), length(m14[2]))
  # We should match as many names as possible
  expect_equal(min(length(m13[1]),length(m13[2])), min(length(sp3[1]),length(sp3[2])) )
  expect_equal(min(length(m14[1]),length(m14[2])), min(length(sp3[1]),length(sp3[2])) )


  m15 <- match.split(sp4)
  m16 <- match.split(sp4, Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m15[1]) + length(m15[2])) <= length(s0))
  expect_true( (length(m16[1]) + length(m16[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m15[1]), length(m15[2]))
  expect_equal(length(m16[1]), length(m16[2]))
  # We should match as many names as possible
  expect_equal(min(length(m15[1]),length(m15[2])), min(length(sp4[1]),length(sp4[2])) )
  expect_equal(min(length(m16[1]),length(m16[2])), min(length(sp4[1]),length(sp4[2])) )

  m17 <- match.split(sp5)
  m18 <- match.split(sp5, Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m17[1]) + length(m17[2])) <= length(s1))
  expect_true( (length(m18[1]) + length(m18[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m17[1]), length(m17[2]))
  expect_equal(length(m18[1]), length(m18[2]))
  # We should match as many names as possible
  expect_equal(min(length(m17[1]),length(m17[2])), min(length(sp5[1]),length(sp5[2])) )
  expect_equal(min(length(m18[1]),length(m18[2])), min(length(sp5[1]),length(sp5[2])) )

  m19 <- match.split(sp6)
  m20 <- match.split(sp6, Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m19[1]) + length(m19[2])) <= length(s0))
  expect_true( (length(m20[1]) + length(m20[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m19[1]), length(m19[2]))
  expect_equal(length(m20[1]), length(m20[2]))
  # We should match as many names as possible
  expect_equal(min(length(m19[1]),length(m19[2])), min(length(sp6[1]),length(sp6[2])) )
  expect_equal(min(length(m20[1]),length(m20[2])), min(length(sp6[1]),length(sp6[2])) )


  m21 <- match.split(sp7)
  m22 <- match.split(sp7, Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m21[1]) + length(m21[2])) <= length(s1))
  expect_true( (length(m22[1]) + length(m22[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m21[1]), length(m21[2]))
  expect_equal(length(m22[1]), length(m22[2]))
  # We should match as many names as possible
  expect_equal(min(length(m21[1]),length(m21[2])), min(length(sp7[1]),length(sp7[2])) )
  expect_equal(min(length(m22[1]),length(m22[2])), min(length(sp7[1]),length(sp7[2])) )


  m23 <- match.split(sp8)
  m24 <- match.split(sp8, Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m23[1]) + length(m23[2])) <= length(s0))
  expect_true( (length(m24[1]) + length(m24[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m23[1]), length(m23[2]))
  expect_equal(length(m24[1]), length(m24[2]))
  # We should match as many names as possible
  expect_equal(min(length(m23[1]),length(m23[2])), min(length(sp8[1]),length(sp8[2])) )
  expect_equal(min(length(m24[1]),length(m24[2])), min(length(sp8[1]),length(sp8[2])) )

  ##### Match based on two subsets

  m25 <- match.pairs(sp1[1],sp1[2])
  m26 <- match.pairs(sp1[1],sp1[2], Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m25[1]) + length(m25[2])) <= length(s0))
  expect_true( (length(m26[1]) + length(m26[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m25[1]), length(m25[2]))
  expect_equal(length(m26[1]), length(m26[2]))
  # We should match as many names as possible
  expect_equal(min(length(m25[1]),length(m25[2])), min(length(sp1[1]),length(sp1[2])) )
  expect_equal(min(length(m26[1]),length(m26[2])), min(length(sp1[1]),length(sp1[2])) )


  m27 <- match.pairs(sp2[1],sp2[2])
  m28 <- match.pairs(sp2[1],sp2[2], Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m27[1]) + length(m27[2])) <= length(s0))
  expect_true( (length(m28[1]) + length(m28[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m27[1]), length(m27[2]))
  expect_equal(length(m28[1]), length(m28[2]))
  # We should match as many names as possible
  expect_equal(min(length(m27[1]),length(m27[2])), min(length(sp2[1]),length(sp2[2])) )
  expect_equal(min(length(m28[1]),length(m28[2])), min(length(sp2[1]),length(sp2[2])) )


  m29 <- match.pairs(sp3[1],sp3[2])
  m30 <- match.pairs(sp3[1],sp3[2], Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m29[1]) + length(m29[2])) <= length(s1))
  expect_true( (length(m30[1]) + length(m30[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m29[1]), length(m29[2]))
  expect_equal(length(m30[1]), length(m30[2]))
  # We should match as many names as possible
  expect_equal(min(length(m29[1]),length(m29[2])), min(length(sp3[1]),length(sp3[2])) )
  expect_equal(min(length(m30[1]),length(m30[2])), min(length(sp3[1]),length(sp3[2])) )


  m31 <- match.pairs(sp4[1],sp4[2])
  m32 <- match.pairs(sp4[1],sp4[2], Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m31[1]) + length(m31[2])) <= length(s0))
  expect_true( (length(m32[1]) + length(m32[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m31[1]), length(m31[2]))
  expect_equal(length(m32[1]), length(m32[2]))
  # We should match as many names as possible
  expect_equal(min(length(m31[1]),length(m31[2])), min(length(sp4[1]),length(sp4[2])) )
  expect_equal(min(length(m32[1]),length(m32[2])), min(length(sp4[1]),length(sp4[2])) )

  m33 <- match.pairs(sp5[1],sp5[2])
  m34 <- match.pairs(sp5[1],sp5[2], Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m33[1]) + length(m33[2])) <= length(s1))
  expect_true( (length(m34[1]) + length(m34[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m33[1]), length(m33[2]))
  expect_equal(length(m34[1]), length(m34[2]))
  # We should match as many names as possible
  expect_equal(min(length(m33[1]),length(m33[2])), min(length(sp5[1]),length(sp5[2])) )
  expect_equal(min(length(m34[1]),length(m34[2])), min(length(sp5[1]),length(sp5[2])) )

  m35 <- match.pairs(sp6[1],sp6[2])
  m36 <- match.pairs(sp6[1],sp6[2], Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m35[1]) + length(m35[2])) <= length(s0))
  expect_true( (length(m36[1]) + length(m36[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m35[1]), length(m35[2]))
  expect_equal(length(m36[1]), length(m36[2]))
  # We should match as many names as possible
  expect_equal(min(length(m35[1]),length(m35[2])), min(length(sp6[1]),length(sp6[2])) )
  expect_equal(min(length(m36[1]),length(m36[2])), min(length(sp6[1]),length(sp6[2])) )


  m37 <- match.pairs(sp7[1],sp7[2])
  m38 <- match.pairs(sp7[1],sp7[2], Competence=10, Intelligence=10)
  # We should not have more than original dataset
  expect_true( (length(m37[1]) + length(m37[2])) <= length(s1))
  expect_true( (length(m38[1]) + length(m38[2])) <= length(s1))
  # Both sets should match in length
  expect_equal(length(m37[1]), length(m37[2]))
  expect_equal(length(m38[1]), length(m38[2]))
  # We should match as many names as possible
  expect_equal(min(length(m37[1]),length(m37[2])), min(length(sp7[1]),length(sp7[2])) )
  expect_equal(min(length(m38[1]),length(m38[2])), min(length(sp7[1]),length(sp7[2])) )


  m39 <- match.pairs(sp8[1],sp8[2])
  m40 <- match.pairs(sp8[1],sp8[3])
  m41 <- match.pairs(sp8[2],sp8[3])
  # We should not have more than original dataset
  expect_true( (length(m39[1]) + length(m39[2])) <= length(s0))
  expect_true( (length(m40[1]) + length(m40[2])) <= length(s0))
  expect_true( (length(m41[1]) + length(m41[2])) <= length(s0))
  # Both sets should match in length
  expect_equal(length(m39[1]), length(m39[2]))
  expect_equal(length(m40[1]), length(m40[2]))
  expect_equal(length(m41[1]), length(m41[2]))
  # We should match as many names as possible
  expect_equal(min(length(m39[1]),length(m39[2])), min(length(sp8[1]),length(sp8[2])) )
  expect_equal(min(length(m40[1]),length(m40[2])), min(length(sp8[1]),length(sp8[3])) )
  expect_equal(min(length(m41[1]),length(m41[2])), min(length(sp8[2]),length(sp8[3])) )
})

test_that("Matches only contain the correct names", {

  ##### Setup

  s0 <- filter.names()
  s1 <- filter.names(Familiarity >= 0.5, Nationality >= 0.5)
  sp1 <- partition.names(Sex)
  sp2 <- partition.names(Sex, discard = 0.2)
  sp3 <- partition.names(Sex, discard = 0.2, subset = s1)
  sp4 <- partition.names.random()
  sp5 <- partition.names.random(subset = s1)
  sp6 <- partition.names.random(prop=c(1,2))
  sp7 <- partition.names.random(subset = s1, prop=c(1,2))
  sp8 <- partition.names.random(prop=c(1,2,4))



  m1 <- match.pairs(sp1[1], sp1[2])

  expect_equal(length(m1[1]), length(m1[1] & sp1[1]))
  expect_equal(length(m1[2]), length(m1[2] & sp1[2]))


  m2 <- match.pairs(sp2[1], sp2[2])

  expect_equal(length(m2[1]), length(m2[1] & sp2[1]))
  expect_equal(length(m2[2]), length(m2[2] & sp2[2]))


  m3 <- match.pairs(sp3[1], sp3[2])

  expect_equal(length(m3[1]), length(m3[1] & sp3[1]))
  expect_equal(length(m3[2]), length(m3[2] & sp3[2]))


  m4 <- match.pairs(sp4[1], sp4[2])

  expect_equal(length(m4[1]), length(m4[1] & sp4[1]))
  expect_equal(length(m4[2]), length(m4[2] & sp4[2]))


  m5 <- match.pairs(sp5[1], sp5[2])

  expect_equal(length(m5[1]), length(m5[1] & sp5[1]))
  expect_equal(length(m5[2]), length(m5[2] & sp5[2]))


  m6 <- match.pairs(sp6[1], sp6[2])

  expect_equal(length(m6[1]), length(m6[1] & sp6[1]))
  expect_equal(length(m6[2]), length(m6[2] & sp6[2]))


  m7 <- match.pairs(sp7[1], sp7[2])

  expect_equal(length(m7[1]), length(m7[1] & sp7[1]))
  expect_equal(length(m7[2]), length(m7[2] & sp7[2]))


  m8 <- match.pairs(sp8[1], sp8[2])
  m9 <- match.pairs(sp8[2], sp8[1])
  m10 <- match.pairs(sp8[1], sp8[3])
  m11 <- match.pairs(sp8[3], sp8[1])
  m12 <- match.pairs(sp8[2], sp8[3])
  m13 <- match.pairs(sp8[3], sp8[2])

  expect_equal(length(m8[1]), length(m8[1] & sp8[1]))
  expect_equal(length(m8[2]), length(m8[2] & sp8[2]))

  expect_equal(length(m9[1]), length(m9[1] & sp8[2]))
  expect_equal(length(m9[2]), length(m9[2] & sp8[1]))

  expect_equal(length(m10[1]), length(m10[1] & sp8[1]))
  expect_equal(length(m10[2]), length(m10[2] & sp8[3]))

  expect_equal(length(m11[1]), length(m11[1] & sp8[3]))
  expect_equal(length(m11[2]), length(m11[2] & sp8[1]))

  expect_equal(length(m12[1]), length(m12[1] & sp8[2]))
  expect_equal(length(m12[2]), length(m12[2] & sp8[3]))

  expect_equal(length(m13[1]), length(m13[1] & sp8[3]))
  expect_equal(length(m13[2]), length(m13[2] & sp8[2]))

})


R/load.R
# Copyright (C) 2018 Tillmann Nett for FernUni Hagen
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
###############################################################################

get.name <- function( name ) {
  names.mean.pc[names.mean.pc$name == name,][-1] %>% unlist()
}


R/split.R
# Copyright (C) 2018 Tillmann Nett for FernUni Hagen
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
###############################################################################

make.split <- function(...) {
  rv <- list( ... )
  class(rv) <- "names.split"
  rv
}

#' Split a set of names into two subsets based on a rating
#'
#' Creates a median split based on a given rating. In addition a
#' percentage of names to close to the median can be discarded.
#' Also allows further splitting already selected subsets of names.
#'
#' @param split       The name of the rating on which the split
#'                    should be performed
#'
#' @param discard     The percentage of names too close to the median
#'                    which should be discarded (default: 0, i.e. keep
#'                    all names)
#'
#' @param subset      An optional subset on which the split should be
#'                    done. If this is left out, the split will
#'                    be created on all names.
#'
#' @return An S3 object of class "names.split". The individual
#'         groups can be retrieved using the [] with the first index
#'         1 for the first group or 2 for the second group.
#'
#' @examples
#'
#' # Split all names along the rating "Sex"
#' s <- partition.names(Sex)
#'
#' # Female names
#' s[1]
#' # Male names
#' s[2]
#'
#' # Same as before, but remove ambigous names (20%)
#' s <- partition.names(Sex, discard=0.2)
#'
#' # Female names
#' s[1]
#' # Male names
#' s[2]
#'
#' # First filter on Competence, then split according to Sex
#' n <- filter.names(Competence >= 0.5)
#' s <- partition.names(Sex, discard=0.2, subset=n)
#'
#' # Female names
#' s[1]
#' # Male names
#' s[2]
#'
#' @importFrom rlang enquo
#' @importFrom rlang !!
#'
#' @export
partition.names <- function(split, discard=0, subset = filter.names()) {

  # We want to make the split work, both if we are given
  # a name, as well as when we are given a character
  split.q <- rlang::enquo( split )

  if( 0 > discard | discard >= 1 ) {
    stop( "Cannot discard ", discard*100, "% of data" )
  }

  # Cutoff values for both groups
  # Centered around the median, with ambigous elements removed
  trgt.high <- 0.5 + discard/2
  trgt.low  <- 0.5 - discard/2

  g1 <- filter.names( !!split.q < trgt.low )
  g2 <- filter.names( !!split.q > trgt.high )

  make.split(low = g1 & subset, high = g2 & subset)
}

#' Split a set of names into random subsets
#'
#' Creates a random split of a subset or all naes. Names can be split
#' into subsets of different size based on the proportion of the names
#' in each subset after the split.
#'
#' @param subset      An optional subset on which the split should be
#'                    done. If this is left out, the split will
#'                    be created on all names.
#'
#' @param prop        Proportion of the names in each subset (defaults to two groups
#'                    of equal size). Proportions are given as a vector with the
#'                    proportion in each group. Proportions are automatically
#'                    normalized.
#'
#' @return An S3 object of class "names.split". The individual
#'         groups can be retrieved using the operator [].
#'
#' @importFrom utils head
#'
#' @examples
#'
#' # Two random groups of equal sizes from all names
#' s <- partition.names.random()
#'
#' # Three groups of different proportions (first group twice as large)
#' s <- partition.names.random( prop = c(2,1,1) )
#'
#' # First filter on Competence, then split randomly
#' n <- filter.names(Competence >= 0.5)
#' s <- partition.names.random(subset=n)
#'
#' @export
partition.names.random <- function(subset = filter.names(), prop = c(0.5,0.5)) {
  idx <- as.numeric( subset )
  idx <- sample(idx)

  c <- seq_along( idx )

  # Normalize proportions
  prop <- (prop / sum(prop)) * length(idx)

  # upper and lower index for each group
  upper <- ceiling( cumsum( prop ) )
  lower <- c(0, head( upper, -1) )

  msks <- lapply(seq_along(upper), function(i) (lower[i] < c) & (c <= upper[i]))
  rv <- lapply(msks, function(msk) make.names.selection( sort( idx[msk] ) ) )
  do.call(make.split, rv)
}

#' @export
`[.names.split` <- function(x, i, j, ...) {
  # No second index, just get the corresponding group
  if( missing(j) ) {
    return( x[[i]] )
  }

  x[[i]][ j ]
}


R/ratings.R
# Copyright (C) 2018 Tillmann Nett for FernUni Hagen
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
###############################################################################

#' Show the ratings for a set of names
#' 
#' @param ...       The ratings that should be retrieved
#' @param subset    The subset for which the ratings should be retrieved
#' 
#' @examples
#' 
#' s <- filter.names(Intelligence > 0.75)
#' 
#' ratings(subset=s, Intelligence, Competence)
#' 
#' @export
ratings <- function(..., subset = filter.names()) {
  qs <- rlang::ensyms( ... )
  if(length(qs) == 0) {
    return( numeric() )
  }
  ratings <- lapply(qs, rlang::eval_tidy, data=names.mean.pc)
  ratings <- lapply(ratings, as.matrix)
  ratings <- do.call(cbind, ratings)
  colnames(ratings) <- lapply(qs, as.character)
  ratings <- ratings[subset,,drop=F]
  rownames(ratings) <-  as.character(subset)
  ratings
}

R/definitions.R
# Copyright (C) 2018 Tillmann Nett for FernUni Hagen
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
###############################################################################

columns.princomp <- c("PC1","PC2","PC3","PC4","PC5")
columns.ratings  <- c( "Education", "Age", "Attractiveness", "Intelligence", "Religiousness", "Familiarity",
                       "Sex", "Nationality", "Modern", "Old", "Ageless", "Competence", "Warmth")

R/filter.R
# Copyright (C) 2018 Tillmann Nett for FernUni Hagen
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License version 3 as
# published by the Free Software Foundation.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
#
###############################################################################

# A selection is a subset of the names, which is stored as an
# index into the complete name set for simplicity.
#
# Since indexes into sets can take different form in R, e.g. a vector
# of logicals which is True for the selecte positions or a vector
# of numericals with the index numbers, we use different representations
# depending on context and convert between them as needed.
# In some cases, this may cause us to loose information, such as duplicates
# or ordering structure, but the operations themselves should usually clearly
# show that this information is lost (e.g. using & on a numeric representation
# looses duplicates, as there is no easy way to keep this information).


make.names.selection <- function(v) {
  rv <- v
  class(rv) <- c("names.selection", class(rv))
  rv
}

#' Filter a subset from the names based on percentiles
#'
#' This method filters a set of names from the complete
#' set. The set is filtered based on predicates, which
#' are evaluated in relation to the percentiles of the
#' given variable (see examples). Multiple predicates
#' are first all evaluated on the complete dataset and
#' then combined (i.e. the percentiles are not
#' re-evaluated in the filtered dataset).
#'
#' @param ... Predicates on the dataset using non-standard evaluation
#'
#' @examples
#'
#' # Retrieve all names without any further reduction
#' filter.names()
#'
#' # Retrieve all names with Intelligence ratings above the median
#' filter.names(Intelligence >= 0.5)
#'
#' # Retrieve all names, which are not too uncommon (Familiarity at least
#' # in the 20th percentile) and for which Intelligence is above the
#' # median
#' filter.names(Familiarity >= 0.2, Intelligence >= 0.5)
#'
#' @importFrom rlang quos eval_tidy
#'
#' @export
filter.names <- function(...) {
  qs   <- rlang::quos( ... )
  if(length(qs) == 0) {
    rv <- rep(T, nrow( names.mean.pc) )
  } else {
    # Create a matrix of percentiles for all attributes
    percs <- seq(0,1,length.out = nrow(names.mean.pc) )
    names.percs <- apply( names.mean.pc[,columns.ratings], MARGIN = 2, function(clm) percs[rank(clm, ties.method="min")])

    msks <- lapply(qs, rlang::eval_tidy, data = as.data.frame( names.percs ) )
    msks <- do.call( cbind, msks )
    rv <- apply(msks, MARGIN=1, all)
  }

  make.names.selection( rv )
}

# Private method for polymorphic lookup depending on representation
names.selection.to.logical <- function(x, ...) {
  UseMethod("names.selection.to.logical")
}

names.selection.to.logical.logical <- function(x, ...) {
  # just strip off class type
  rv <- x
  class(rv) <- "logical"
  rv
}

names.selection.to.logical.numeric <- function(x, ...) {
  rv <- rep(F, nrow(names.mean.pc))
  rv[x] <- T
  rv
}

#' @export
as.logical.names.selection <- function(x, ...) {
  names.selection.to.logical(x, ...)
}

# Private method for polymorphic lookup depending on representation
names.selection.to.double <- function(x, ...) {
  UseMethod("names.selection.to.double")
}

names.selection.to.double.logical <- function(x, ...) {
  rv <- as.double( which( x ) )
  rv
}

names.selection.to.double.numeric <- function(x, ...) {
  # Just strip of class type
  rv <- x
  class(rv) <- "numeric"
  rv
}

#' @export
as.double.names.selection <- function(x, ...) {
  names.selection.to.double(x, ...)
}



#' @export
as.character.names.selection <- function(x, ...) {
  # We can select here independently of the actual representations
  # so we do not need any conversion
  names.mean.pc[x, "name"]
}

#' @export
as.data.frame.names.selection <- function(x, ...) {
  ratings <- rlang::ensyms( ... )
  if( length( ratings ) == 0 ) {
    return( data.frame( name = as.character(x), stringsAsFactors = F ) )
  }
  data.frame( name = as.character(x), ratings(subset=x, ...), row.names = seq_along(x), stringsAsFactors = F )
}

#' @export
print.names.selection <- function(x, ... ) {
  print( as.data.frame( x, ...) )
}

#' @export
`&.names.selection` <- function(x,y) {
  # To intersect two selections, we must ignore the ordering
  # and multiples, so we always have to convert to logical
  # first
  rv <- as.logical( x ) & as.logical( y )
  class(rv) <- c("names.selection",class(rv))
  rv
}

#' @export
`[.names.selection` <- function(x, i, ...) {
  # If we select, we may always have multiples
  # so we must always return a numeric index
  #
  # Convert to numeric index first, so we know
  # that we can just select correctly

  x.numeric <- as.numeric( x )
  # Select the correct elements
  rv <- x.numeric[ i ]
  class(rv) <- c("names.selection", class(rv))
  rv
}

# Private helper method to use polymorphism on the internal representation
the.length <- function(x) {
  UseMethod("the.length")
}

the.length.logical <- function(x) {
  sum(x)
}


the.length.numeric <- function(x) {
  length( as.numeric( x ) )
}

#' @export
length.names.selection <- function(x) {
  the.length(x)
}
